### PR TITLE
Add support for `js-unknown-ghcjs` -- for 19.03

### PR DESCRIPTION
**Pull request by angerman**
This adds enough logic to nixpkgs to support the `js-unknown-ghcjs` triple.

###### Motivation for this change

Support `--target=js-unknown-ghcjs` where supported.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Ericson2314

### CODE REVIEW

File: lib/systems/examples.nix [hunk 1]
```diff
@@ -204,4 +204,10 @@ rec {
     libc = "msvcrt"; # This distinguishes the mingw (non posix) toolchain
     platform = {};
   };
+
+  # Ghcjs
+  ghcjs = {
+    config = "js-unknown-none-ghcjs";
```
Comment (matthewbauer) [inline lib/systems/examples.nix]: I'm not sure about this. While I agree that we need better cross-compilation support for ghcjs, "js-unknown-none-ghcjs" is not a widely agreed upon name. [Searching for js-unknown-none-ghcjs](https://www.google.com/search?q=js-unknown-none-ghcjs) only gives me #65589. I was going to recommend using "asmjs-unknown-emscripten" which is used by Rust as well, but then realized that GHCJS is not technically asmjs. In addition, we have no C compiler that is compatible with GHCJS, so things like bootstrapping are not well defined.

Comment (matthewbauer) [reply lib/systems/examples.nix]: I think it's okay to allow this one though. GHCJS is in a pretty unique situation, so there is not that much downside to allowing it on an experimental basis. I would like to get some agreement with @luite and ghcjs maintainers on this though.

Comment (matthewbauer) [reply lib/systems/examples.nix]: I would suggest using ```js-unknown-ghcjs``` though. ```none-ghcjs``` doesn't really make sense. The "environment" is supposed to qualify the "os", but that doesn't make sense when your os is none. For instance try this:

```
$ ./config.sub asmjs-unknown-none-ghcjs
Invalid configuration `asmjs-unknown-none-ghcjs': system `none-ghcjs' not recognized
```

Comment (angerman) [reply lib/systems/examples.nix]: @matthewbauer thanks for the review. I agree that `js-unknown-ghcjs` is better. The primary reason for choosing that is that it's already pretty close to how cabal thinks about ghcjs.

We (that includes @luite) will move ghcjs more towards a proper ghc cross compiler. @Ericson2314 will be able to supply you with some more background knowledge from ICFP I guess.

@luite *is* working on some emscripten support for ghcjs, so we may eventually interact with emscripten, however that may only work from within cabal anyway due to calling convention needs and as such never really be the same as asmjs targets. If we can make that work eventually we may need to add another target.

Note that this needs to be fixed in Nixpkgs, as we can not locally fix it via overlays, as per #65589.

input-output-hk/haskell.nix's overlay branch (input-output-hk/haskell.nix#210) will allow you to treat ghcjs like a regular cross compiler as in
```
nix build '(with import ./. { nixpkgs = ../nixpkgs; nixpkgsArgs = { crossSystem = { config = "js-unknown-ghcjs"; }; }; }; (haskell-nix.hackage-package { name = "hello"; version = "1.0.0.2"; })).components.exes.hello'
```
after removing a few more minor kinks.

**Comment by angerman** [comment]
After some additional lengthy discussion with @luite, we'll try to use `asmjs-unknown-emscripten` and see if we can make that work.

**Comment by Ericson2314** [comment]
Err but aren't we outside the asmjs subset?

**Comment by angerman** [comment]
@Ericson2314, morally I think `js-unknown-ghcjs` is the most correct. @luite is experimenting with the emscripten toolchain for c parts in libraries. Now we could reach for `js-ghcjs-emscripten` I believe. @luite made the argument that `js` is not a wildly accepted architecture in autotools and `asmjs` is.

When choosing between `wasm32/wasm` and `asmjs`, i believe we are closer to `asmjs` in spirit than to `wasm` :-/

I’ll have to experiment a bit more with more complex libraries today to get a better understanding of the benefits of each triple.

I welcome any ideas!

**Comment by Ericson2314** [comment]
I think let's do `js-unknown-ghcjs` or `js-unknown-none` then. If/when the emscripten part is done we can switch, but this is most correct for the GHCJS that is released today. (I can't say I'm wild about enscrypten is its such a big hack that will definitely increase the number of things to worry about and maintain.) I wouldn't worry about what GNU Config does; I rewrote large swaths of it (http://git.savannah.gnu.org/cgit/config.git/log/?qt=author&q=John+Ericson) and am on good terms with the maintainer wrt upstreaming things.

CC @matthewbauer?

**Comment by angerman** [comment]
ok, let's do `js-unknown-ghcjs` for now, and switch (or add more) if we end up having it working. `js-unknown-ghcjs` at least works for simple haskell packages for me already.

**Comment by angerman** [comment]
Alright, can we get this merged then?

File: pkgs/development/libraries/libiconv/default.nix [hunk 2]
```diff
@@ -20,7 +20,7 @@ stdenv.mkDerivation rec {
   ];
 
   postPatch =
-    lib.optionalString ((stdenv.hostPlatform != stdenv.buildPlatform && stdenv.hostPlatform.libc == "msvcrt") || stdenv.cc.nativeLibc)
+    lib.optionalString ((stdenv.hostPlatform != stdenv.buildPlatform && stdenv.hostPlatform.libc == "msvcrt") || (stdenv.cc != null && stdenv.cc.nativeLibc))
```
Comment (Ericson2314) [inline pkgs/development/libraries/libiconv/default.nix]: I'm going to merge without that for now, but we can get back to it tomorrow / in a few hours.

**Pull request merged**